### PR TITLE
PD-486: SideNav Component

### PR DESCRIPTION
**Pull request by seanmhanson**
## ✍️ Proposed Changes and Additions

As part of IA Navigation, one major goal outside of the new MongoNav is to unify the side navigation appearance and behavior for most of our Cloud products, including:
- Atlas & Cloud Manager (Atlas, Cloud Manager, Organizations, Projects, User Preferences)
- Ops Manager 4.4 (Ops Manager, Organizations, Projects, User Account)
- MongoDB Account
- Federations / IDPs
- Stitch/Realm
- Charts

This component establishes the style and pattern we will use across all applications. It allows individual items to be customized to use specific components, allowing integration with React Router and other components.

One benefit of adding this component early is that we can update all of Cloud's navigation immediately (save for the removal of Stitch and Charts links), since it telegraphs no upcoming changes and is functionally independent of the rest of IA. Get pumped.

🎟 _Jira ticket:_ [PD-486](https://jira.mongodb.org/browse/PD-486)

## 🛠 Types of Changes
- [x] Tooling (updates to workspace config or internal tooling)
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

I have yet to complete the documentation, and some automatically generated documentation is not yet present but should be easy to add.

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

## 📷 Demo

Screen Shot 2020-01-24 at 12 33 11 PM

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,6 +15,13 @@ storybook/public/
 # JUnit report files
 reports
 
+# OS generated files
+.DS_Store
+.DS_Store?
+.Trashes
+ehthumbs.db
+Thumbs.db
+
```
Comment (seanmhanson) [inline .gitignore]: This was a drive-by, because I was seeing macOS things try to sneak into my commits, and the thumbs.db tend to sneak in similarly when I am on my desktop back home.

File: packages/menu/src/MenuItem.tsx [hunk 2]
```diff
@@ -184,6 +189,7 @@ const MenuItem = React.forwardRef(
       disabled = false,
       active = false,
       className,
+      titleTextClassName,
```
Comment (seanmhanson) [inline packages/menu/src/MenuItem.tsx]: Before this is complete, I need to examine this, because for some reason in some cases this is showing up as an attribute `titletextclassname` on the link/button, rather than as part of the class composition by Emotion.

File: packages/side-nav/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Side Nav
```
Comment (seanmhanson) [inline packages/side-nav/README.md]: (to be added when next possible)

File: packages/side-nav/src/SideNav.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    sideNavGroup: 'side-nav-group',
+    sideNavItem: 'side-nav-item',
+    sideNavLink: 'side-nav-link',
+  };
+
+  let renderedEls: RenderedElements = {};
+
+  afterEach(() => {
+    document.body.innerHTML = '';
+    renderedEls = {};
```
Comment (seanmhanson) [inline packages/side-nav/src/SideNav.spec.tsx]: [nitpick] I can remove the `afterAll` and add a call to `cleanup` here since this `afterEach` is at the root.

File: packages/side-nav/src/SideNav.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+      beforeEach(() => {
+        const { sideNavLink } = testIds;
+        const children = (
+          <a href="#clusters" data-testid={sideNavLink}>
+            Clusters
+          </a>
+        );
+
+        renderSideNavItem({ children });
+      });
```
Comment (seanmhanson) [inline packages/side-nav/src/SideNav.spec.tsx]: [TODO] this lacks a test to detect the correctly rendered child (this was a brief oversight, it should be simple)

File: packages/side-nav/src/SideNav.story.tsx [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import { SideNav, SideNavItem, SideNavGroup } from './index';
+
+function MockAtlasSideNav() {
+  return (
+    <SideNav>
+      <SideNavGroup headerText="Data Storage">
+        <SideNavItem href="#clusters" active>
+          Clusters (active)
+        </SideNavItem>
+        {/* <SideNavItem>Triggers (not a link)</SideNavItem>
```
Comment (seanmhanson) [inline packages/side-nav/src/SideNav.story.tsx]: These were commented out when testing something, and I accidentally committed the comments. These will be re-enabled.

File: packages/side-nav/src/SideNavGroup.tsx [hunk 7]
```diff
@@ -0,0 +1,82 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { css, cx } from '@leafygreen-ui/emotion';
+import { uiColors } from '@leafygreen-ui/palette';
+import { createDataProp } from '@leafygreen-ui/lib';
+import { ulStyleOverrides, LEFT_RIGHT_OFFSET } from './styles';
+
+export const sideNavGroupDataProp = createDataProp('side-nav-group-section');
```
Comment (seanmhanson) [inline packages/side-nav/src/SideNavGroup.tsx]: [question] I don't think this is actually being used, but I saw it in other components. Can this safely be removed?

File: packages/side-nav/src/SideNavGroup.tsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+      )}
+      <ul className={ulStyleOverrides}>{children}</ul>
+    </li>
+  );
+}
+
+SideNavGroup.displayName = 'SideNavGroup';
+
+SideNavGroup.propTypes = {
+  children: PropTypes.node,
```
Comment (seanmhanson) [inline packages/side-nav/src/SideNavGroup.tsx]: [correction needed] `children: PropTypes.node.isRequired`

File: packages/side-nav/src/SideNav.tsx [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+      <ul className={ulStyleOverrides}>{children}</ul>
+    </nav>
+  );
+}
+
+SideNav.displayName = 'Side Nav';
+
+SideNav.propTypes = {
+  className: PropTypes.string,
+  children: PropTypes.node,
```
Comment (seanmhanson) [inline packages/side-nav/src/SideNav.tsx]: [correction needed] `children: PropTypes.node.isRequired`

File: packages/side-nav/src/SideNavItem.tsx [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+SideNavItem.displayName = 'SideNavItem';
+
+SideNavItem.propTypes = {
+  active: PropTypes.bool,
+  disabled: PropTypes.bool,
+  className: PropTypes.string,
+  description: PropTypes.string,
+  children: PropTypes.node,
```
Comment (seanmhanson) [inline packages/side-nav/src/SideNavItem.tsx]: [correction needed] `children: PropTypes.node.isRequired`

File: .changeset/tiny-lies-type.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
+---
+'@leafygreen-ui/menu': minor
+'@leafygreen-ui/side-nav': minor
```
Comment (brookescarlett) [inline .changeset/tiny-lies-type.md]: new components should be a major bump so we can get to 1.0.0 when we release; unless we are planning to release this as an alpha/beta like we are doing with MongoNav

File: packages/side-nav/package.json [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    "types": "./dist/index.d.ts",
+    "scripts": {
+      "build": "../../node_modules/.bin/webpack --config ../../webpack.config.js"
+    },
+    "license": "Apache-2.0",
+    "publishConfig": {
+      "access": "public"
+    },
+    "devDependencies": {
+      "@leafygreen-ui/lib": "^3.2.0"
```
Comment (brookescarlett) [inline packages/side-nav/package.json]: I have to update this in the boilerplate, but please use lib at 4.0.0

File: packages/side-nav/src/SideNav.spec.tsx [hunk 13]
```diff
@@ -0,0 +1,224 @@
+import React from 'react';
+import '@testing-library/jest-dom/extend-expect';
+import { render, cleanup } from '@testing-library/react';
+import { MenuItemProps } from '@leafygreen-ui/menu';
+import { SideNav, SideNavGroup, SideNavItem } from './index';
+
+type renderedElement = HTMLElement | null | undefined;
```
Comment (brookescarlett) [inline packages/side-nav/src/SideNav.spec.tsx]: Because you are implementing this type as optional, undefined is already inferred for you so you shouldn't need tot explicitly define here

File: packages/side-nav/src/SideNav.story.tsx (hunk 6 above)
Comment (brookescarlett) [reply packages/side-nav/src/SideNav.story.tsx]: Can add knobs here, so that users can play with active/non-active, href/nohref and the text of the SideNavItem

File: packages/side-nav/src/SideNav.tsx [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+      aria-label="side-nav"
+      className={cx(sideNavStyles, className)}
+      {...rest}
+    >
+      <ul className={ulStyleOverrides}>{children}</ul>
+    </nav>
+  );
+}
+
+SideNav.displayName = 'Side Nav';
```
Comment (brookescarlett) [inline packages/side-nav/src/SideNav.tsx]: [nit]: value of displayName should be one word: SideNav

File: packages/side-nav/src/SideNavGroup.tsx (hunk 7 above)
Comment (brookescarlett) [reply packages/side-nav/src/SideNavGroup.tsx]: yes this can be removed if you aren't using the sideNavGroup selector, this prop is used to give us more specificity with css styles since we can't reference exact classNames

File: packages/side-nav/src/SideNavGroup.tsx [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+ */
+function SideNavGroup({
+  children,
+  className,
+  headerText,
+  headerClassName,
+  ...rest
+}: SideNavGroupProps) {
+  return (
+    <li className={className} {...rest} {...sideNavGroupDataProp.prop}>
```
Comment (brookescarlett) [inline packages/side-nav/src/SideNavGroup.tsx]: might need to add aria-level props to the list items (on this level and the one below), see Accessibility tab on storybook for more details

Comment (brookescarlett) [reply packages/side-nav/src/SideNavGroup.tsx]: actually seems to be complaining about lack of menu/menu bar labels, which is interesting because wai-aria doesn't implement those roles on it's own side nav... something to consider

File: packages/side-nav/src/styles.ts [hunk 16]
```diff
@@ -0,0 +1,11 @@
+import { css } from '@leafygreen-ui/emotion';
+
+export const LEFT_RIGHT_OFFSET = 15;
```
Comment (brookescarlett) [inline packages/side-nav/src/styles.ts]: 👍

File: packages/side-nav/package.json (hunk 12 above)
Comment (brookescarlett) [reply packages/side-nav/package.json]: also add the rest of the devDeps (menu, palette, emotion, etc.)

File: .gitignore (hunk 1 above)
Comment (DesignerDave) [reply .gitignore]: Hmm... So far we've avoided adding editor / OS-specific files like these, and instead have each person use their own global gitignore to avoid files specific to their environment from being added. That said, I don't feel strongly about this.

@hswolff Do you feel strongly about it?

File: packages/menu/src/MenuItem.tsx (hunk 2 above)
Comment (DesignerDave) [reply packages/menu/src/MenuItem.tsx]: Oh weird. I don't see anything particularly wrong here with that 🤷‍♂

File: packages/menu/src/MenuItem.tsx [hunk 17]
```diff
@@ -219,18 +225,22 @@ const MenuItem = React.forwardRef(
...
+            )}
           >
             {children}
           </div>
           {description && (
             <div
               className={cx(descriptionTextStyle, {
                 [activeDescriptionTextStyle]: active,
-                [disbaledTextStyle]: disabled,
+                [disabledTextStyle]: disabled,
```
Comment (DesignerDave) [inline packages/menu/src/MenuItem.tsx]: I'd coordinate with @brookescarlett around changes to this file in particular. I know she also has a pull request in where this is fixed.

File: packages/side-nav/src/SideNavGroup.tsx [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+  </SideNavItem>
+</SideNavGroup>
+ * ```
+ * @param props.children Content that will appear inside of SideNavGroup component.
+ * @param props.className ClassName that will be applied to root SideNavGroup element.
+ * @param props.headerText Header text displayed above the SideNavGroup contents.
+ * @param props.headerClassName ClassName that will be applied to SideNavGroup's header element.
+ *
+ */
+function SideNavGroup({
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavGroup.tsx]: Just want to get @brookescarlett to weigh in here – from an accessibility standpoint, this seems correct overall to me, but I could be wrong.

File: packages/side-nav/src/SideNavGroup.tsx [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+  children,
+  className,
+  headerText,
+  headerClassName,
+  ...rest
+}: SideNavGroupProps) {
+  return (
+    <li className={className} {...rest}>
+      {headerText && (
+        <div className={cx(sideNavLabelStyle, headerClassName)} role="heading">
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavGroup.tsx]: We should probably use a more semantic tag than `div` here. Perhaps something like `label` or `h4`?

Comment (DesignerDave) [reply packages/side-nav/src/SideNavGroup.tsx]: Additionally, I think we get `role="header"` for free that way.

File: packages/side-nav/src/SideNavGroup.tsx [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+ */
+function SideNavGroup({
+  children,
+  className,
+  headerText,
+  headerClassName,
+  ...rest
+}: SideNavGroupProps) {
+  return (
+    <li className={className} {...rest}>
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavGroup.tsx]: Perhaps I'm missing something, but we can probably pass className via `rest` here.

File: packages/side-nav/src/SideNavGroup.tsx [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+  font-weight: bold;
+  text-transform: uppercase;
+  padding: 5px ${LEFT_RIGHT_OFFSET}px 5px ${LEFT_RIGHT_OFFSET}px;
+`;
+
+interface SideNavGroupProps {
+  /**
+   * Content that will appear inside of SideNavGroup component.
+   */
+  children: React.ReactNode;
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavGroup.tsx]: Is there a reason this is a required prop?

File: packages/side-nav/src/SideNavGroup.tsx [hunk 22]
```diff
@@ -0,0 +1,79 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { css, cx } from '@leafygreen-ui/emotion';
+import { uiColors } from '@leafygreen-ui/palette';
+import { ulStyleOverrides, LEFT_RIGHT_OFFSET } from './styles';
+
+const sideNavLabelStyle = css`
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavGroup.tsx]: If we already have styles split from the component, should this class live in the styles file instead?

File: packages/side-nav/src/SideNavGroup.tsx [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+  display: block;
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavGroup.tsx]: We should get this for free so long as we're using a div or header. This is fine to keep if we switch to a label tag I believe.

File: packages/side-nav/src/SideNavItem.tsx [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+const containerStyleOverrides = css`
+  padding: 10px ${LEFT_RIGHT_OFFSET}px 10px ${LEFT_RIGHT_OFFSET}px;
+  min-height: 0;
+  border-radius: 5px;
+
+  &:before {
+    content: normal;
+  }
+
+  &:focus:active {
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavItem.tsx]: It would be nice to implement this focus state conditionally with the [useUsingKeyboardContext](https://github.com/mongodb/leafygreen-ui/tree/master/packages/leafygreen-provider#useusingkeyboardcontext) hook from the leafygreen-provider package.

This way applications with the provider can benefit from better focus state handling (and it gracefully degrades when the provider isn't a parent).

File: packages/side-nav/src/SideNavItem.tsx [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+  }
+`;
+
+const titleTextStyleOverrides = css`
+  font-size: 15px;
+  text-transform: capitalize;
+`;
+
+const disabledStyleOverrides = css`
+  background-color: ${uiColors.white};
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavItem.tsx]: We should probably set to transparent (assuming the intent is for it to not have a visible background). Additionally, we probably should change the color to a lighter gray to make disabled states more apparent.

File: packages/side-nav/src/SideNav.tsx [hunk 26]
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { css, cx } from '@leafygreen-ui/emotion';
+import { ulStyleOverrides } from './styles';
+
+const sideNavStyles = css`
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNav.tsx]: You might consider inlining this single-liner to improve readability (not that it's terrible now).

File: packages/side-nav/src/SideNav.tsx [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+  width: 200px;
+`;
+
+interface SideNavProps {
+  className?: string;
+  children: React.ReactNode;
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNav.tsx]: The props are currently pretty simple, but probably worth adding TSDoc annotations here for a consistent DX.

File: packages/side-nav/src/SideNavItem.tsx [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+    content: normal;
+  }
+
+  &:focus:active {
+    background-color: ${uiColors.gray.light2};
+  }
+`;
+
+const titleTextStyleOverrides = css`
+  font-size: 15px;
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavItem.tsx]: This is a larger font size than in the prototype

File: packages/side-nav/src/SideNavItem.tsx [hunk 29]
```diff
@@ -0,0 +1,79 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { css, cx } from '@leafygreen-ui/emotion';
+import { uiColors } from '@leafygreen-ui/palette';
+import { MenuItem, MenuItemProps } from '@leafygreen-ui/menu';
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavItem.tsx]: So there's one main reason I think we might consider not using MenuItem as the base here: We lose a lot of control over the transition between states of the MenuItem.

I think with the design of these, we have a great micro-interaction opportunity in animating the scale of the background between different states, but (barring some hacky implementation), those animations will be difficult to add without full control of the rendered DOM.

Additionally, there's the added benefit of not needing to modify MenuItem for this component which is nice for two reasons:
1. We don't need to expose a className prop for MenuItem, which we want to avoid style overrides on in general.
2. There won't be conflicts with Brooke's work.

Open to thoughts on this, of course.

File: packages/side-nav/src/SideNavItem.tsx [hunk 30]
```diff
@@ -0,0 +1,79 @@
...
+import { LEFT_RIGHT_OFFSET } from './styles';
+
+// style overrides for menu item
+const containerStyleOverrides = css`
+  padding: 10px ${LEFT_RIGHT_OFFSET}px 10px ${LEFT_RIGHT_OFFSET}px;
+  min-height: 0;
+  border-radius: 5px;
+
+  &:before {
+    content: normal;
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavItem.tsx]: Is this to unset the left bar in the MenuItem? I think `initial` is preferable, if so.

File: packages/side-nav/package.json (hunk 12 above)
Comment (DesignerDave) [reply packages/side-nav/package.json]: @brookescarlett Is there a reason it's in the boilerplate? I don't think every component uses this.

Comment (brookescarlett) [reply packages/side-nav/package.json]: ^ yes, but can make a separate PR quickly to update

File: packages/side-nav/src/SideNavGroup.tsx (hunk 19 above)
Comment (brookescarlett) [reply packages/side-nav/src/SideNavGroup.tsx]: :+1 @DesignerDave

File: .gitignore (hunk 1 above)
Comment (hswolff) [reply .gitignore]: Nope, looks grrrreat!

File: packages/side-nav/tsconfig.json [hunk 31]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (hswolff) [inline packages/side-nav/tsconfig.json]: We need to add a reference to this project in the root `tscnofig.json` file so that when we run `yarn ts` from the root it picks up and properly type checks this package.

File: .changeset/two-mirrors-glow.md [hunk 32]
```diff
@@ -0,0 +1,6 @@
+---
+'@leafygreen-ui/side-nav': major
+'@leafygreen-ui/menu': minor
+---
+
+Mongo Menu now adds a class to style title text. Side Nav has been introduced as a new component.
```
Comment (hswolff) [inline .changeset/two-mirrors-glow.md]:
```suggestion
MongoMenu now adds a class to style title text. SideNav has been introduced as a new component.
```

To mirror the component names themselves?

File: packages/side-nav/src/SideNavItem.tsx [hunk 33]
```diff
@@ -0,0 +1,79 @@
...
+  disabled = false,
+  className,
+  titleTextClassName,
+  children,
+  description,
+  href,
+  ...rest
+}: MenuItemProps) {
+  if (active) {
+    rest['aria-current'] = 'page';
```
Comment (hswolff) [inline packages/side-nav/src/SideNavItem.tsx]: Shouldn't be modifying the props value.

Comment (seanmhanson) [reply packages/side-nav/src/SideNavItem.tsx]: I landed on a better solution that is more usable as well.

File: packages/side-nav/src/SideNavGroup.tsx (hunk 19 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNavGroup.tsx]: I wanted to use a h tag, but didn't know what level you all use for what. H4 it is!

File: packages/side-nav/src/SideNavGroup.tsx (hunk 20 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNavGroup.tsx]: Oh gosh, I completely forgot about that and just pulled everything out by habit

File: packages/side-nav/src/SideNavGroup.tsx (hunk 21 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNavGroup.tsx]: Nope, I think it can be optional, if you want to use a self closing tab. However, what would that mean in terms of the UI? A group with no members behind a label?

File: packages/side-nav/src/SideNavGroup.tsx (hunk 22 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNavGroup.tsx]: I only split out common styles (anything reused). Whenever possible otherwise, I try to keep styles alongside components.

File: packages/side-nav/src/SideNav.tsx (hunk 27 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNav.tsx]: Yup, part of my current work is consistency amongst all forms of docs

File: packages/side-nav/src/SideNavGroup.tsx (hunk 19 above)
Comment (DesignerDave) [reply packages/side-nav/src/SideNavGroup.tsx]: I'm sure we'll have stricter guidelines in the future, but for the time being this should do!

File: packages/side-nav/src/SideNavGroup.tsx (hunk 21 above)
Comment (DesignerDave) [reply packages/side-nav/src/SideNavGroup.tsx]: In theory – I could go either way here. I suppose there isn't a lot of importance to typing it either way, just that it's perfectly valid to not pass children.

File: packages/side-nav/src/SideNavGroup.tsx (hunk 22 above)
Comment (DesignerDave) [reply packages/side-nav/src/SideNavGroup.tsx]: That's fair enough!

File: packages/side-nav/src/SideNavGroup.tsx (hunk 23 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNavGroup.tsx]: H4 it is, so we'll get this and bold for free, but pay for it in margin top and bottom.

File: packages/side-nav/src/SideNavItem.tsx (hunk 25 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNavItem.tsx]: The lighter color is already taken care of by the default styles for `MenuItem`, but I've updated the background.

File: packages/side-nav/src/SideNavItem.tsx (hunk 30 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNavItem.tsx]: It is meant to unset the pseudo-elements, but MDN specifically states that this is the intended usage of `normal` (see [refs here](https://developer.mozilla.org/en-US/docs/Web/CSS/content#values))

File: packages/side-nav/src/SideNavItem.tsx (hunk 28 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNavItem.tsx]: Updated to 14px.

File: packages/side-nav/src/SideNavItem.tsx (hunk 30 above)
Comment (DesignerDave) [reply packages/side-nav/src/SideNavItem.tsx]: This is more of a preference, but usually I use `initial` rather than the default value to clearly communicate the intent of the rule to set it to its default value. Without looking at documentation, that's unclear here unless you're intimately familiar with the content property.

That said, I don't feel strongly about this. Either value should function the same here.

Comment (seanmhanson) [reply packages/side-nav/src/SideNavItem.tsx]: This will be addressed as part of removing MenuItem as a dependency for SideNavItem

File: packages/menu/src/MenuItem.tsx (hunk 17 above)
Comment (seanmhanson) [reply packages/menu/src/MenuItem.tsx]: This will be addressed as part of removing MenuItem as a dependency for SideNavItem. I'll revert the changes I made beyond pulling out the repeated types.

File: packages/menu/src/MenuItem.tsx (hunk 2 above)
Comment (seanmhanson) [reply packages/menu/src/MenuItem.tsx]: This will be addressed as part of removing MenuItem as a dependency for SideNavItem, so no worries

File: packages/lib/src/index.ts [hunk 34]
```diff
@@ -59,3 +59,24 @@ export const keyMap = {
   Space: 32,
   Tab: 9,
 };
+
+/**
+ * An enum of accepted values for the "aria-current" attribute, used for
```
Comment (brookescarlett) [inline packages/lib/src/index.ts]: 👍

File: packages/side-nav/README.md [hunk 35]
```diff
@@ -0,0 +1,171 @@
...
+
+![npm (scoped)](https://img.shields.io/npm/v/@leafygreen-ui/side-nav.svg)
+
+## Example
+
+```js
+import { SideNav, SideNavGroup, SideNavItem } from '@leafygreen-ui/sideNav';
```
Comment (brookescarlett) [inline packages/side-nav/README.md]: the package name should be side-nav

File: packages/side-nav/src/SideNav.tsx [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+  </SideNavGroup>
+</SideNav>
+ * ```
+ * @param props.className Class name that will be applied to SideNav root component.
+ * @param props.children Content that will appear inside of SideNav root component.
+ *
+ */
+function SideNav({ children, ...rest }: SideNavProps) {
+  return (
+    <nav style={{ width: '200px' }} aria-label="side-nav" {...rest}>
```
Comment (brookescarlett) [inline packages/side-nav/src/SideNav.tsx]: can we move this to a className with our emotion instance

File: packages/side-nav/src/SideNavItem.tsx [hunk 37]
```diff
@@ -0,0 +1,285 @@
...
+  const { usingKeyboard: showFocus } = useUsingKeyboardContext();
+
+  let Root: ElementType<any>;
+
+  if (usesLinkElement(props)) {
+    Root = RootComponentTypes.Link;
+  } else if (usesCustomElement(props)) {
+    Root = props[RootComponentTypes.Custom];
+  } else {
+    Root = RootComponentTypes.Div;
```
Comment (brookescarlett) [inline packages/side-nav/src/SideNavItem.tsx]: I think we should probably default to a button here. Since this is a navigation component, it's fair to assume that most, if not all, of the items should respond to a click.

File: packages/side-nav/src/SideNav.tsx (hunk 36 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNav.tsx]: Haha, @DesignerDave asked the opposite and said to inline it, so I got rid of that. I actually like to have it as a class instead and it's mostly a nitpick, so I'm going to revert it back. Sorry Dave!

Comment (brookescarlett) [reply packages/side-nav/src/SideNav.tsx]: I think maybe he meant `<nav className={css`width: 200px`} />` but perhaps I'm wrong. Either way definitely support using className rather than style 👍

File: packages/menu/src/MenuItem.tsx [hunk 38]
```diff
@@ -143,12 +143,10 @@ interface SharedMenuItemProps {
    * Class name that will be applied to root MenuItem element.
    */
   className?: string;
-
```
Comment (hswolff) [inline packages/menu/src/MenuItem.tsx]: nit, I prefer the extra white space between the definitions as it gives space for my eyes to rest before reading the next prop name.

Having no line break makes it feel like a run-on sentence.

File: packages/side-nav/package.json [hunk 39]
```diff
@@ -0,0 +1,22 @@
+
+  {
+    "name": "@leafygreen-ui/side-nav",
+    "version": "1.0.0",
```
Comment (hswolff) [inline packages/side-nav/package.json]: nit, I believe we have to have this be initialized to 0.9.0 so that the changelog script correctly runs the first time we use it. Gonna let @brookescarlett weigh in.

File: packages/menu/src/MenuItem.tsx (hunk 38 above)
Comment (hswolff) [reply packages/menu/src/MenuItem.tsx]: That being said, I don't feel strong enough to desire a code change, but just adding my opinion into the mix.

File: packages/side-nav/src/SideNavGroup.tsx [hunk 40]
```diff
@@ -0,0 +1,78 @@
...
+ */
+function SideNavGroup({
+  headerText,
+  headerClassName,
+  children,
+  ...rest
+}: SideNavGroupProps) {
+  return (
+    <li {...rest}>
+      {headerText && (
```
Comment (hswolff) [inline packages/side-nav/src/SideNavGroup.tsx]: In an effort to mildly premature optimize.....I wonder if we should rename this prop to `header` and allow it to to be any valid react node type. If it's a string it does what's here, but if it's not then it just renders itself.

Thinking is instead of us someday adding a `tooltipNode` prop in the future we just expose that additional control. 

Happy to be pushed back on this, @DesignerDave usually does :).

File: packages/side-nav/src/index.ts [hunk 41]
```diff
@@ -0,0 +1,5 @@
+import SideNav from './SideNav';
+import SideNavItem, { SideNavItemProps } from './SideNavItem';
+import SideNavGroup from './SideNavGroup';
+
+export { SideNav, SideNavItem, SideNavGroup, SideNavItemProps };
```
Comment (hswolff) [inline packages/side-nav/src/index.ts]: I'm wondering if rather than add this to our public export API we just let the test import the props directly from the module.

File: packages/side-nav/src/SideNav.spec.tsx [hunk 42]
```diff
@@ -0,0 +1,286 @@
...
+
+  let renderedEls: RenderedElements = {};
+
+  afterEach(() => {
+    document.body.innerHTML = '';
+    renderedEls = {};
+    cleanup();
+  });
+
+  describe('Side Nav Item', () => {
```
Comment (hswolff) [inline packages/side-nav/src/SideNav.spec.tsx]: Let's have these mirror the component names. Should make grepping easier.

File: packages/menu/src/MenuItem.tsx (hunk 38 above)
Comment (seanmhanson) [reply packages/menu/src/MenuItem.tsx]: Definitely agree actually, and I've made those changes. Things changed as a result of copy/pasting when trying to make sure I used the same descriptions across the typescript, the docstring, and the readme.

File: packages/side-nav/src/index.ts (hunk 41 above)
Comment (seanmhanson) [reply packages/side-nav/src/index.ts]: Clever. I didn't think of using two import statements like that.

File: packages/side-nav/src/SideNavGroup.tsx (hunk 40 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNavGroup.tsx]: I wanted to do this, and opted not to because of premature optimization concerns when I was trying to get this out the door. I agree though and am going to change this. It also simplifies our props, as we don't need to take a custom class anymore when you can just directly pass in the header.

I will have to think a little bit about this and ensuring there is either an `h*` tag or `role="header"` somewhere.

**Comment by seanmhanson** [comment]
@hswolff RFAL if you continue to sneakily do reviews on a Saturday

File: packages/side-nav/src/SideNavGroup.tsx [hunk 43]
```diff
@@ -42,23 +41,24 @@ interface SideNavGroupProps {
...
- *
  */
-function SideNavGroup({
-  headerText,
-  headerClassName,
-  children,
-  ...rest
-}: SideNavGroupProps) {
+function SideNavGroup({ header, children, ...rest }: SideNavGroupProps) {
+  const renderHeader = () => {
```
Comment (hswolff) [inline packages/side-nav/src/SideNavGroup.tsx]: micro opt, rather than incurring a new function call cost we could do:

```jsx
const renderedHeader = typeof header === 'string' ? <h4 className={cx(sideNavLabelStyle)}>{header}</h4> : header;
```

Could even inline it.

File: packages/menu/src/MenuItem.tsx [hunk 44]
```diff
@@ -88,7 +88,7 @@ const titleTextStyle = css`
   }
 `;
 
-const activetitleTextStyle = css`
+const activeTitleTextStyle = css`
```
Comment (DesignerDave) [inline packages/menu/src/MenuItem.tsx]: Brooke fixed both of the typos in this file in her PR. Let's omit these changes so we keep the concerns separated.

File: .changeset/clever-dolls-know.md [hunk 45]
```diff
@@ -0,0 +1,7 @@
+---
+'@leafygreen-ui/side-nav': major
+'@leafygreen-ui/lib': minor
+'@leafygreen-ui/menu': patch
```
Comment (DesignerDave) [inline .changeset/clever-dolls-know.md]: As noted below, let's omit this altogether to keep from any conflicts with Brooke's PR.

File: packages/side-nav/src/SideNavGroup.tsx [hunk 46]
```diff
@@ -0,0 +1,78 @@
...
+ *   If a string is provided, it will be rendered with default styling as a header tag.
+ * @param props.children Class name that will be applied to the component's header.
+ */
+function SideNavGroup({ header, children, ...rest }: SideNavGroupProps) {
+  return (
+    <li {...rest}>
+      {typeof header === 'string' ? (
+        <h4 className={sideNavLabelStyle}>{header}</h4>
+      ) : (
+        header
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavGroup.tsx]: Hmm... I have mixed feelings about not rendering the H4 when the prop isn't a string. Reason being is that if anyone needed to include an icon or span in the header, they'd have to duplicate the header styles entirely...

File: packages/menu/src/MenuItem.tsx (hunk 44 above)
Comment (seanmhanson) [reply packages/menu/src/MenuItem.tsx]: This was a complete oversight (ditto the mention of it in the changeset file). Thanks for noticing this.

File: packages/side-nav/src/SideNavGroup.tsx (hunk 46 above)
Comment (seanmhanson) [reply packages/side-nav/src/SideNavGroup.tsx]: Definitely, good plan. Theoretically if someone wants a custom header without those styles, they should just pass it in as part of `children`.

File: packages/side-nav/src/SideNavItem.tsx [hunk 47]
```diff
@@ -0,0 +1,301 @@
...
+  padding: 10px ${LEFT_RIGHT_OFFSET}px 10px ${LEFT_RIGHT_OFFSET}px;
+  min-height: 0;
+  display: flex;
+  flex-direction: column;
+  border: none;
+  border-radius: 5px;
+  cursor: pointer;
+  text-align: left;
+  font-family: Akzidenz, ‘Helvetica Neue’, Helvetica, Arial, sans-serif;
+  transition: background-color 150ms ease-in-out;
```
Comment (DesignerDave) [inline packages/side-nav/src/SideNavItem.tsx]: So I was trying to figure out why the scale transition doesn't work between states... We shouldn't care about background color on this element. The background color should be applied to the `:before` pseudo-class instead.
